optimum/habana/transformers/models/llama/modeling_llama.py

```diff
@@ -31,9 +31,9 @@
 
 def update(prev, cur, dim, idx):
     orig_cur = cur
-    if prev.shape == cur.shape:
+    if cur.shape[2] > 1 and cur.shape[2] <= prev.shape[2]:
         # Initialize
-        prev.copy_(cur)
+        prev[:, :, :idx, :].copy_(cur)
```

Comment:
@puneeshkhanna what is the real shape of cur, I think the shape of prev is [N, M, max_T, H]

Reply:
See below change which gets executed only when past key values is None or basically for the 1st token.
cur will be set to the actual seq len for the 1st token [N, M, actual seq len, H]

elif reuse_cache and token_idx is not None:
            # With reuse_cache, KV cache is pre allocated hence for the 1st token we can slice the inputs till token idx for the fwd pass
            input_ids = input_ids[:, :token_idx]
            attention_mask = attention_mask[:, :token_idx]

Reply:
so, if length of prompts between different BS are not same, there will be recompilation for 1st token?
correct me if I'm wrong

Reply:
That is always the case irrespective of my change and even recreation of KV cache. 
I m pushing a separate PR to have an addition argument of max input tokens before hand to avoid such recompilations.

Reply:
if with max_input, here we can not use idx, we should use max_input_length right?

Reply:
token_idx is always set to the index of the token that needs to be generated. With max input length, tokenizer will pad/truncate the input seq till that max input tokens. For that, change is only in run generation script only.

Reply:
https://github.com/huggingface/optimum-habana/pull/426 - PR for configurable max input tokens to avoid graph recompiles between batches of various lengths.

Reply:
@ZhaiFeiyue - I think I m understanding the concern since token idx may change between batches. Let me try it out once.

Reply:
@ZhaiFeiyue - we are fine. Preferably we should set max input tokens and since padding is always done on left side, token_idx will be always set to the max input token value only during the 1st token generation for any batch of input sequences and hence no recompilations.

Reply:
![image](https://github.com/huggingface/optimum-habana/assets/80079571/25ee5a84-26c3-4563-a7b0-276146ffecbd)

see above picture, token_idx is not equal to max_input_length I think

Reply:
what I think is here,  prev[:, :, :max_input_length, :].copy_(cur)
if user not set max_input_length, I think it should be max_token_length, since with static shape the length of cur is max_token_length

Reply:
No the green padding is on the left side only for decoder text generation otherwise generation won't be correct.  token_idx initialization below to input_tensor.shape[-1].
Am I missing anything ?

# token_idx is the current index in the generation process, it is incremented each time a new token is generated
model_kwargs["token_idx"] = torch.tensor(inputs_tensor.shape[-1], device=inputs_tensor.device)

Reply:
For example see dump of input tokens of BS-4 below with max input tokens as say 50 ; padding is always on the left side.

'input_ids': tensor([[    0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     1,  7569,   650,   338,  9796,
           322,   306,   508,  7569,   650,   338,  9796,   322,   306,   508],
        [    0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     1,   450,   716, 14064,   393,  2355, 19054,
           445,  1629,   450,   716, 14064,   393,  2355, 19054,   445,  1629],
        [    0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     1,
           512,   278,  2215,  2215,  5418,   515,  1749, 15400, 29891, 29892,
           512,   278,  2215,  2215,  5418,   515,  1749, 15400, 29891, 29892],
        [    0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     1,
         24819,   338,   278,   871,   982, 24819,   338,   278,   871,   982,
         24819,   338,   278,   871,   982, 24819,   338,   278,   871,   982]])

Reply:
thanks @puneeshkhanna, seems that current implement of `max_input_tokens` is different from what I think.
if with the current implement, you are right.
I will merge